### DISCUSSION TITLE
Wrapping keras optimizer with hvd optimizer breaks the code

### DISCUSSION THREAD

**Discussion by bluepra** [horovod/horovod#3624] [Q&A] [answered]
Using 2 VMs with 1 GPU each to do distributed training using Horovod.

This is what I have for building and compiling my model:
![image](https://user-images.githubusercontent.com/71447892/181047510-73c55456-190f-46a9-9e1b-84cf14438e7f.png)

When I pass in True for horovod, my code breaks starts Epoch 1, then hangs for a while (no output), and then crashes with the following error.
![image](https://user-images.githubusercontent.com/71447892/181049061-e47ebd10-786a-469a-bdf4-aff289864142.png)
![image](https://user-images.githubusercontent.com/71447892/181049272-55cdb0c0-4267-4448-b781-28702f2138b0.png)

When I pass in False for horovod, my code works on 2 VMs.

My hvd.init() and model.fit() are both in a script called horovod_training.py.
And my build_and_compile_model() function is in another script called model.py. 
In model.py I have the following imports:
![image](https://user-images.githubusercontent.com/71447892/181049736-6ae849d2-8f6b-4692-bbd1-d385d5d12c8e.png)

Is this okay to do? Any help is appreciated!

**Comment by maxhgerlach** [comment]
Do the Keras examples from the Horovod repo work for you? https://github.com/horovod/horovod/tree/master/examples/keras 

Those examples use `Adadelta`. Do they still work for you if you replace it with `Adam`?

**Comment by bluepra** [comment]
I tested it with https://github.com/horovod/horovod/blob/master/examples/tensorflow2/tensorflow2_keras_mnist.py and it worked. This code was using tf.optimizers.Adam(scaled_lr).

**Comment by maxhgerlach** [comment]
Let's move this to Discussion as the problem is probably specific to your script.

**Reply by bluepra** [reply]
Ok would you like to see some specific part of my code?

**Comment by bluepra** [comment]
@maxhgerlach 
Can I please get a reply on this discussion?

**Reply by maxhgerlach** [reply]
I don't know your code and I wouldn't have time to study it. In your screenshot it looks like different ranks try to submit different tensors to the allreduce at the same time, thus blocking each other. That won't work.

**Reply by bluepra** [reply]
So how do you recommend I go about debugging this?

**Comment by bluepra** [comment, marked as answer]
So I think I got it working. I was using the tensorboard callback and only adding that to my chief worker, this was causing some rank issues. Now every worker gets the tensorboard callback, and the horovod training seems to run just fine!
